### PR TITLE
Correct use of 'nul' 'null' and capitalization in the book

### PR DESCRIPTION
**Pull request by alexandermerritt**
r? @steveklabnik

### CODE REVIEW

**Comment by steveklabnik** [comment]
Hey @alexandermerritt ! Sorry this review took a while.

So, we don't have any standard here, but IIRC, @retep998 specifically used `NUL` in those contexts because they were meaningful. Maybe they can chime in and confirm?

**Comment by retep998** [comment]
I use `NUL` when referring to the DOS device equivalent to `/dev/null`.

File: src/doc/book/ffi.md [hunk 1]
```diff
@@ -521,14 +521,14 @@ against `libc` and `libm` by default.
 
 # The "nullable pointer optimization"
 
-Certain types are defined to not be `null`. This includes references (`&T`,
+Certain types are defined to not be `NULL`. This includes references (`&T`,
```
Comment (retep998) [inline src/doc/book/ffi.md]: Why is it back ticked here? It's definitely not an identifier that Rust uses, so it seems wrong.

Comment (steveklabnik) [inline src/doc/book/ffi.md]: Yeah, I would also think that this shouldn't be backticked

**Comment by retep998** [comment]
The way [MSDN](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365247(v=vs.85).aspx) does it is they refer to strings as being null terminated, while the ascii character itself is referred to as `NUL` yet they call it the null character.

**Comment by alexandermerritt** [comment]
@steveklabnik @retep998 Thanks for the feedback!

The use of these terms seems to not be fully consistent and I believe leads to confusion. For example, [Wikipedia](https://en.wikipedia.org/wiki/Null-terminated_string) uses "Null-terminated string" to mean C string terminated with ```'\0'``` as does your reference in MSDN, but MSDN also calls the ending byte in a C string as the "null character". I know them as ```NUL``` being the single-byte character marking the end of a C string ```'\0'``` (as worded in [POSIX](http://man7.org/linux/man-pages/man3/strncat.3p.html)), and ```NULL``` being the value of a pointer which is invalid or uninitialized (in [POSIX](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/stddef.h.html) but also referred to as the "null pointer").

I do not agree in written text that both a string can be "null-terminated" when there exists also a "null pointer", as a "null string" then means what? (A "C string" ends with ```NUL``` but can have value ```NULL``` which means it then does not exist.)

My philosophy is, when referring to the _values_, ```NUL``` as ```'\0'``` and ```NULL``` as ```(void*)0```, they are capitalized. For example, stating "compare to null" is misleading; does it mean compare a single byte, or a word size? Thus strings are terminated with ```NUL``` or "```NUL```-terminated" and invalid pointers are ```NULL``` or are "null pointers".

Perhaps most of this is just a writing style preference? My motivation is that many people are still confused as to the difference between ```NUL```, ```NULL``` (wrt. their size, and their context), and "null", so I wanted to help standardize their use, but please feel free to choose what you feel is appropriate.

**Comment by steveklabnik** [comment]
Whew, even in a language without null, null is a problem :sweat:

**Comment by AndrewBrinker** [comment]
A small comment: would it be useful to start some Rust project language guide to document these sorts of practices? That way there's something to point new contributors toward when writing docs, and obvious against which existing documentation can be evaluated.

**Pull request merged**